src/map/homunculus.c

```diff
@@ -1177,138 +1178,173 @@ static bool homunculus_shuffle(struct homun_data *hd)
+	if (libconfig->setting_lookup_int(it, "EvoId", &i32) != CONFIG_TRUE || !homdb_checkid(i32)) {
+		ShowError("homunculus_read_db_libconfig_sub: Invalid homunculus EvoId (%d) provided for entry %d in '%s', skipping...\n", i32, idx, source);
+		return false;
 	}
+	db->evo_class = i32;
 
+	if (libconfig->setting_lookup_string(it, "Name", &str) != CONFIG_TRUE) {
+		ShowError("homunculus_read_db_libconfig_sub: Invalid homunculus Name '%s' provided for entry %d in '%s', skipping...\n", str, idx, source);
+		return false;
+	}
+	safestrncpy(db->name, str, NAME_LENGTH - 1);
+
+	if (!itemdb->lookup_const(it, "FoodItem", &i32)) {
+		ShowError("homunculus_read_db_libconfig_sub: Invalid homunculus FoodItem '%d' provided for entry %d in '%s', skipping...\n", i32, idx, source);
+		return false;
+	}
+	db->foodID = i32;
+
+	if (libconfig->setting_lookup_int(it, "HungryDelay", &i32) == CONFIG_TRUE && i32 >= 0)
+		db->hungryDelay = i32;
+	if (mob->lookup_const(it, "Size", &i32) && i32 >= 0)
+		db->base_size = i32;
+	if (mob->lookup_const(it, "EvoSize", &i32) && i32 >= 0)
+		db->evo_size = i32;
+	if (mob->lookup_const(it, "Race", &i32) && i32 >= 0)
+		db->race = i32;
+	if (mob->lookup_const(it, "Element", &i32) && i32 >= 0)
+		db->element = i32;
+	if (libconfig->setting_lookup_int(it, "Aspd", &i32) == CONFIG_TRUE && i32 >= 0)
+		db->baseASPD = i32;
+
+	if (!homun->read_db_libconfig_sub_stats(it, idx)) {
+		ShowError("homunculus_read_db_libconfig_sub: Failed to read homunculus stats for entry %d in '%s', skipping...\n", idx, source);
+		return false;
+	}
+
+	return true;
+}
+
+static bool homunculus_read_db_libconfig_sub_stats(struct config_setting_t *it, int idx)
+{
+	nullpo_retr(false, it);
+	Assert_retr(false, idx >= 0 && idx < MAX_HOMUNCULUS_CLASS);
+
+	struct s_homunculus_db *db = &homun->dbs->db[idx];
+	struct config_setting_t *t = NULL;
+
+	if ((t = libconfig->setting_get_member(it, "BaseStats")) != NULL && config_setting_is_group(t))
+		homun->read_db_libconfig_sub_stats_group(t, &db->base, NULL);
+	if ((t = libconfig->setting_get_member(it, "GrowthStats")) != NULL && config_setting_is_group(t))
+		homun->read_db_libconfig_sub_stats_group(t, &db->gmin, &db->gmax);
+	if ((t = libconfig->setting_get_member(it, "EvolutionStats")) != NULL && config_setting_is_group(t))
+		homun->read_db_libconfig_sub_stats_group(t, &db->emin, &db->emax);
+
+	return true;
 }
+
+static bool homunculus_read_db_libconfig_sub_stats_group(struct config_setting_t *it, struct h_stats *smin, struct h_stats *smax)
+{
+	nullpo_retr(false, it);
+	nullpo_retr(false, smin);
+
+
+	if (smax != NULL) {
+		struct config_setting_t *t = NULL;
+
+#define homunculus_read_stats_minmax(name, min, max) \
+	do { \
+		if ((t = libconfig->setting_get_member(it, name)) != NULL && config_setting_is_list(t)) { \
+			min = libconfig->setting_get_int_elem(t, 0); \
+			max = libconfig->setting_get_int_elem(t, 1); \
+			if (min > max) \
+				min = max; \
```

Comment:
Suggesting some sanitization of the macro:

```suggestion
#define homunculus_read_stats_minmax(name, min, max) \
	do { \
		struct config_setting_t *t = NULL; \
		if ((t = libconfig->setting_get_member(it, (name))) != NULL && config_setting_is_list(t)) { \
			(min) = libconfig->setting_get_int_elem(t, 0); \
			(max) = libconfig->setting_get_int_elem(t, 1); \
			if ((min) > (max)) \
				(min) = (max); \
```